### PR TITLE
Hermione and cucumber custom steps

### PR DESCRIPTION
**Pull request by epszaw**
### Context

Provides some useful updates for `hermione` and `cucumber` reporters.

The biggest change here - implementation for custom steps with can be used with any framework integration.

#### Checklist
- [x] [Sign Allure CLA][cla]
- [x] Provide unit tests

[cla]: https://cla-assistant.io/accept/allure-framework/allure2

### CODE REVIEW

File: packages/allure-js-commons/src/AllureCommandStep.ts [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+
+  /**
+   * Recursively writes attachments from the given step and all it's children
+   * Mutates given step object!
+   */
+  static writeStepAttachments(runtime: AllureRuntime, step: ExecutableItem) {
+    if (step.attachments.length > 0) {
+      step.attachments.forEach((attachment) => {
+        const encoding = /(text|application)/.test(attachment.type) ? "utf8" : "base64";
+        const attachmentContent = Buffer.from(attachment.source, encoding);
```
Comment (vovsemenv) [inline packages/allure-js-commons/src/AllureCommandStep.ts:82]: video to base64 sounds really exprensive and long 🤔 did we really need it?

File: packages/allure-js-commons/src/AllureCommandStep.ts [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+      environmentInfo,
+    );
+    this.metadata.steps = (this.metadata.steps || []).concat(steps);
+  }
+
+  async start(body: StepBodyFunction): Promise<AttachmentMetadata> {
+    const startDate = new Date().getTime();
+
+    try {
+      const res = body.call(this, this);
```
Comment (vovsemenv) [inline packages/allure-js-commons/src/AllureCommandStep.ts:218]: this start will be executed on the reporter thread for the paralell execution??? if so something can broke

File: packages/allure-js-commons/src/AllureCommandStep.ts (hunk 1 above)
Comment (epszaw) [reply packages/allure-js-commons/src/AllureCommandStep.ts:82]: We can't send any binary data as buffer using ipc text messages. Could you propose something here?

File: packages/allure-js-commons/src/AllureCommandStep.ts (hunk 2 above)
Comment (epszaw) [reply packages/allure-js-commons/src/AllureCommandStep.ts:218]: It depends. For example `wdio` attaches separated report for each thread. 
Cucumberjs doesn't, but it executes steps inside world methods, which call in separated threads too.
What kind of doubt you have?

File: packages/allure-hermione/src/index.ts [hunk 3]
```diff
@@ -235,6 +278,25 @@ const hermioneAllureReporter = (hermione: HermioneAllure, opts: AllureReportOpti
     browser.addCommand("attach", async (testId: string, source: string, mimetype: string) => {
       await addAttachment(testId, source, mimetype);
     });
+    browser.addCommand("category", async (testId: string, category: Category) => {
+      await sendMetadata(testId, {
```
[the remark this answers is not in the document]
Comment (epszaw) [reply packages/allure-hermione/src/index.ts:282]: We do the same in cucumber js reporter. All the environment info and categories will be accumulated and written as single file.

**Review by vovsemenv — APPROVED** [review]
Everything works just like a cake 🎉
Tested with cucumber  and cucumber + pw combinations

**Review by baev — APPROVED** [review]

**Pull request merged**